src/java.base/share/classes/com/sun/crypto/provider/DHPublicKey.java

```diff
@@ -106,7 +223,7 @@ final class DHPublicKey implements PublicKey,
         try {
             this.key = new DerValue(DerValue.tag_Integer,
                                     this.y.toByteArray()).toByteArray();
-            this.encodedKey = getEncoded();
+            this.encodedKey = encode(p, g, l, key);
         } catch (IOException e) {
             throw new ProviderException("Cannot produce ASN.1 encoding", e);
         }
```

Comment:
You should remove the try-catch. `encode` doesn't throw IOE. This also removes the diff to jdk 21.

Reply:
It is toByteArray that trows the IOException.
I can add the InternalError here as in the other places?